### PR TITLE
[Backport 0.23] Replicate snapshot faster

### LINKED ISSUE #5135: Restart takes too long or never completes when snapshot contains many files
**Describe the bug**

Snapshot replication initiated via raft sends the chunks sequentially. So if the snapshot has many chunks it takes too long to replicate. Since this often happens during restart, a follower takes several minutes to catch up its state. If this takes longer than the snapshot period, the leader might start a new replication. As a result the node restart never completes.  
Related #4887

To fix this we should improve the snapshot replication logic in raft to sends more chunks in parallel.

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

Backport of #5189 

## Related issues

closes #5135

### CODE REVIEW

**Comment by deepthidevaki** [comment]
image

**Review by ChrisKujawa — APPROVED** [review]
Uuuh yeah :D thanks @deepthidevaki good work.

The changes look similar to the other branch, not really a difference or?

**Comment by deepthidevaki** [comment]
Yes. It's the same changes, but since the code base has changed it is not a simple cherry-pick. Also the test is not backported.

File: atomix/cluster/src/main/java/io/atomix/raft/roles/AbstractAppender.java [hunk 1]
```diff
@@ -447,45 +449,63 @@ protected void handleConfigureResponseError(
...
+            snapshot.getPath(),
+            e);
+        return Optional.empty();
+      }
       member.setNextSnapshotIndex(snapshot.index());
       member.setNextSnapshotChunk(null);
     }
 
     final InstallRequest request;
-    synchronized (snapshot) {
```
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/roles/AbstractAppender.java:459]: I removed this `synchronized`. I think it is not required. What do you think? @Zelldon

Comment (Zelldon) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/AbstractAppender.java:459]: I think the same, since we having only one thread.

Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/AbstractAppender.java:459]: :+1:

**Pull request merged**